### DISCUSSION TITLE
Teensy Board 2.09, is WIFI possible?

### DISCUSSION THREAD

**Discussion by Team-Deepflights** [terjeio/grblHAL#341] [General]
As title says, I would like to send my code over Wifi, has anyone done this?

I have a few Wemos R1/D1 here which could be used, but I would need some help (changing firmware to ? and so on..) 

Thanks in advance :-)

**Comment by terjeio** [comment]
I have D1 minis that I intend to check out - but when? There surely must projects out there that use these cards to bridge raw Telnet or Websocket comms to RS-232/UART.
If not I believe grblHAL has most of the bits needed - lwIP Telnet and Websocked code and UART code for ESP32 that may be reused? The code has to be split out and glued together to make it work.

**Comment by Team-Deepflights** [comment]
That was my thought, a Wifi/UART bridge.
I saw in my machine.h `#define BLUETOOTH_ENABLE    1 // Set to 1 for HC-05 module. Requires and claims one auxillary input pin.`
Is that explained somewhere what this module can do and why an AUX in pin is needed? If BT work, it will be easy to use a D1 instead.
Control the lathe or mill as I can do with the IO Sender would be really nice, but for now sending the g-code would be absolutely enough.

> If not I believe grblHAL has most of the bits needed - lwIP Telnet and Websocked code and UART code for ESP32 that may be reused? The code has to be split out and glued together to make it work.

This is also good point, let me try the first one, if this don't work, I'll go to the second option. :-)

**Reply by terjeio** [reply]
> Is that explained somewhere what this module can do and why an AUX in pin is needed?

A description can be found in the [plugin readme](https://github.com/grblHAL/Plugins_Bluetooth).

The aux pin is used when auto configuring and to automatically switch to streaming data via Bluetooth when a client connects. On disconnect streaming falls back to USB.

> Control the lathe or mill as I can do with the IO Sender would be really nice, but for now sending the g-code would be absolutely enough.

Control via senders is what it is about, when configured it shows up as a normal COM port.

**Comment by Team-Deepflights** [comment]
First of all, it works! :-)
After try&error with the machine.h I am able to connect over Telnet and able to open IOSender over it.
I would like to contribute and share the way to do it, how can we realize it? I think in here it would dissapear sooner or later.

**Reply by phil-barrett** [reply]
Best to put in the wiki?  If you write it up, I can add it. 

One suggestion, can you also do it with websockets?  That has been a more reliable connection method.

I'd also like to do a blog on this.  Definitely worth having available. I will, of course, fully credit you.

**Reply by terjeio** [reply]
How did you make it work? By using an existing project or creating a new?

> One suggestion, can you also do it with websockets? That has been a more reliable connection method.

What is the issue with the telnet connection? If speed then this was improved lately.
The iMXRT1062 is insane speed wise, in check mode I get transfer speeds around 350Kbytes/s - nearly 5 times faster than MSP432.

**Reply by phil-barrett** [reply]
> What is the issue with the telnet connection? If speed then this was improved lately.

There was a recently fixed bug relating to lwip on Teensys. Not sure everyone trying it out will have the latest source code and where there is smoke, there may be fire.

**Reply by Team-Deepflights** [reply]
> > What is the issue with the telnet connection? If speed then this was improved lately.
> 
> There was a recently fixed bug relating to lwip on Teensys. Not sure everyone trying it out will have the latest source code and where there is smoke, there may be fire.

I have the latest source code, everthing is OK

**Comment by Team-Deepflights** [comment]
Wiki is the right place, i'll write it as PDF as send it to you and you also can use it for your blog of course.
You mean do the connection over port 80 or 81? I try.. :-)

**Comment by Team-Deepflights** [comment]
The PDF is done.
Have fun :-)

@Phil and Terjeio:
I you need the original screenshots for futher manuals or blogs, let me know.
It was a great pleasure for me to contribute a little bit to this exceptional project!

[Guide to add Wifi on BOARD.pdf](https://github.com/terjeio/grblHAL/files/7782068/Guide.to.add.Wifi.on.BOARD.pdf)

**Reply by phil-barrett** [reply]
Thanks.  Are you using an actual Wemos D1 mini or one of the many clones out there?  Are they all basically alike?

Also, do you have explicit wiring diagrams available?

**Reply by terjeio** [reply]
I have tested with an unbranded D1 mini:

Be aware that the UART port is used for data transfer, by default this is limited to 115200 baud. It can be increased to 460800...

In Check mode: a 200K gcode file transfers in 27s at 115200, 12s at 460800. Via telnet over ethernet - less than 1 second (a 4.8Mb file transfers in 13s). So not very useful for high speed laser engraving. 

If the card could be configured to act as an USB host I guess the transfer speed could be increased dramatically.

From the guide:

> To use the Telnet port we need to reconfigure the board, open your maschine.h file and uncomment the following:
//#define NETWORK_TELNET_PORT 23 at line 107

This is not correct, it is for enabling telnet over ethernet. Comment out this line instead to use UART:
`#define USB_SERIAL_CDC      2 // 1 for Arduino class library and 2 for PJRC C library. Comment out to use UART communication.`

AFAIKT websocket comms is not supported by the firmware. If it could be modified to change a pin depending on client connected or not automatic switching between USB and UART could be added, the Bluetooth plugin does that.

**Comment by Team-Deepflights** [comment]
I have both, clones and Wemos D1, all are the same.
I have no wiring diagram, the pins you need are printed on the PCB of the D1 Mini, I suppose that 4 wires isn't really complicated, but of course it's described it in the PDF. 
A wiring diagramm in Fritzing or similar would be nice.

**Comment by phil-barrett** [comment]
OK, I just ordered several D1 minis. Will use this as an opportunity to get up on this.  I'm sure I will have some questions.

**Reply by phil-barrett** [reply]
Hi, my first question (to both Team-Deepflights and Terje) - Would it be possible to program the ESP-12F via the teensy rather than though the USB-Serial adaptor? Even better, is it possible to program the ESP-12F via a plug-in and not require the end user to deal with it at all? Are the programming protocols open source?

**Comment by Team-Deepflights** [comment]
It's not a rocket science, really easy to do, but I you have some questions, I am here :-)

**Comment by Team-Deepflights** [comment]
All you need to get the Wifi to work is open source, I don‘t know if it‘s easy to implement the code, it would be a nice have to, but to be honest, the upload of the firmware I described is not that hard.

Btw, I have now ESP 01S module running up, way more smaller.
The Esp-Link software work there too.
As they don‘t have USB, best way to program them is with a dedicaded programmer.

The links are from Ebay Germany, but it helps to get the hardware wherever you are.

ESP 01S module:

https://www.ebay.de/itm/5-x-ESP-01S-Wifi-Modul-ESP8266-WLAN-Wireless-Serial-Board-ESP01-ESP-01-Arduino-/164761463134?mkcid=16&mkevt=1&_trksid=p2349624.m46890.l49286&mkrid=707-127634-2357-0

Programmer:

https://www.ebay.de/itm/1-5-x-ESP-01-ESP-01S-USB-Programmer-mit-Schalter-Modul-ESP8266-ESP01-UART-/164772618358?mkcid=16&mkevt=1&_trksid=p2349624.m46890.l49286&mkrid=707-127634-2357-0

I think this is the best choice, as the pins of the PCB are are different to the pinout I needed, I simply wired them.
Maibe for another version of the PCB it would be nice to rework the pins for the ESP 01S?

**Reply by phil-barrett** [reply]
I'm thinking if I just use an ESP-12F it will be pretty small and don't need any of the D1 circuitry if I can power it directly with 3.3V.  Just need to program it. Probably will need a carrier to adapt to the serial pins on the breakout board - it will be only slightly larger than the ESP-12F. Antenna issues need to get worked through.  The ESP-01 could be smaller using it's TH footprint, though.

**Comment by Team-Deepflights** [comment]
It was my first tought, take a Wemos board to program it and desolder it after programming it.
But this is the reason why I did not use it, if I want to upload another firmware I have to desolder it and solder it again into the breakout board.
The ESP01 is simply easier in handling.

Btw, it would be really a nice feature to write to the SD card :-)
I noticed that if I upload a NC file and close the ioSender windows and open it with another Laptop (grbl HAL board still running over external 5v) the nc file upped is gone.
Or is an EEPROM the solution?

After a few days using the board over wifi I have to say the everything runs as smooth as directly connected over USB

**Reply by terjeio** [reply]
> I noticed that if I upload a NC file and close the ioSender windows and open it with another Laptop (grbl HAL board still running over external 5v) the nc file upped is gone.

Uploaded to a SD card or just loaded into ioSender? If uploaded to a card it should still be there.

> Or is an EEPROM the solution?

For storing gcode programs in the controller then no - EEPROMS do not come in large enough sizes for that.

**Comment by Team-Deepflights** [comment]
That's why I wrote that writing on SD would be a nice feature, I did not manage to save the code on SD.
Is that only possible over FTP?

**Reply by terjeio** [reply]
> Is that only possible over FTP?

You can try with `#define SDCARD_ENABLE 2` in _my_machine.h_, this will enable upload via the YModem protocol. I did not manage to make this work in a 100% reliable fashion so I have kept it secret...

**Comment by Team-Deepflights** [comment]
Ahhh. enable 2 :-)
Thx, I try that later.

**Comment by Team-Deepflights** [comment]
Hm, I'm also not happy with it.
Upload is OK, running the file the ioSender crashes.
Deleting the it crashes too.
Maibe one of the needed libraries is wrong or too new.
I finish the lathe first, this is a secondary theater ;-)
